### PR TITLE
Add container linux update operator

### LINKED ISSUE #2080: Provide and configure locksmithd for CoreOS instances
**User Story**
As a cluster user who deploys multiple CoreOS instances, I'd like them to reboot one by one upon a CoreOS update being released, instead of (worst case) all of them at once.

**Acceptance criteria**
[ ] either locksmithd or coreos-update-operator are configured on all CoreOS nodes in user clusters
[ ] the configuration enforces only one node to be rebooted at a time

### LINKED ISSUE #2996: Add ContainerLinuxUpdateOperator
**User Story**
As a user of Kubermatic i want that my ContainerLinux based cluster does coordinated reboots.

The [ContainerLinuxUpdateOperator](https://github.com/coreos/container-linux-update-operator) does exactly that.

Issue though is that user clusters can have mixed nodes, which will make the pods fail if they run on a non ContainerLinux node.

To schedule the pods only on ContainerLinux nodes they need to be labeled. (By default there is no label on Nodes which would enable this)

**Acceptance criteria**
- [ ] Extend the user cluster controller to label all ContainerLinux nodes with the label `kubernetes.io/coreos=true`
- [ ] Add a addon for the ContainerLinuxUpdateOperator

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
This PR adds the ContainerLinux controller which is responsible for:
- Labeling all Nodes which run ContainerLinux
- Deploy the ContainerLinuxUpdateOperator

When there is no ContainerLinux node in the cluster, the operator won't be deployed.
All operator componentes get deployed with a NodeSelector, so they only run on ContainerLinux nodes.

If the user removes all ContainerLinux nodes, the operator will not be removed - I don't think its worth the effort (Rare edge case + It does not hurt to have those resources).

Fixes #2996
Fixes #2080

**Release note**:
```release-note
Add the ContainerLinuxUpdateOperator to all clusters using ContainerLinux nodes
```

### CODE REVIEW

File: api/pkg/controller/container-linux/container_linux_controller.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	events := make(chan event.GenericEvent)
+	src := &source.Channel{Source: events, DestBufferSize: 1}
+	if err := c.Watch(src, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+
+	timer := time.NewTimer(syncInterval)
+	go func() {
+		<-timer.C
+		events <- event.GenericEvent{}
```
Comment (mrIncompetent) [inline api/pkg/controller/container-linux/container_linux_controller.go]: I really did not wanted to sync on each node event.

Syncing on time might not be the cleanest way - but IMO its an acceptable trade-off.
Current interval is `30s`

File: api/pkg/controller/container-linux/container_linux_controller.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	if err := c.Watch(src, &handler.Funcs{GenericFunc: func(event event.GenericEvent, queue workqueue.RateLimitingInterface) {
+		queue.Add(reconcile.Request{})
+	}}); err != nil {
+		return err
+	}
+
+	// Don't react on node events - There are just too many node events happening
+	// Instead just make sure we create a Watcher for nodes, so we get the cache
+	noOpHandler := &handler.Funcs{}
+	return c.Watch(&source.Kind{Type: &corev1.Node{}}, noOpHandler)
```
Comment (alvaroaleman) [inline api/pkg/controller/container-linux/container_linux_controller.go]: Why not simply use a predicate function that ingores update events where `oldNode.Status.NodeInfo.OperatingSystem == newNode.Status.NodeInfo.OperatingSystem` , somewhat like this https://github.com/kubermatic/zevenet-lb-controller/blob/cbab0452694f590d7513231c0a22505512b2af7a/pkg/controller/service/service_controller.go#L95 ?

File: api/pkg/controller/container-linux/resources/clusterrole.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/kubermatic/kubermatic/api/pkg/resources/reconciling"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+const (
+	ClusterRoleName = "container_linux_update_operator"
+)
+
+func ClusterRoleCreator() reconciling.NamedClusterRoleCreatorGetter {
```
Comment (alvaroaleman) [inline api/pkg/controller/container-linux/resources/clusterrole.go:13]: Soomer or late we should write a CLI that can generate this code from a YAML document....

**Comment by alvaroaleman** [comment]
Please make the PR title a bit more meaningful (`Add container linux update operator`?). 

More of a sidequestion, does the operator drain and taint nodes prior to upgrading them?

File: api/pkg/controller/container-linux/container_linux_controller.go (hunk 2 above)
Comment (mrIncompetent) [reply api/pkg/controller/container-linux/container_linux_controller.go]: I've added a update predicate func to check if the node is using ContainerLinux + missing the label.

That should cover it :)

File: api/pkg/controller/container-linux/resources/clusterrole.go (hunk 3 above)
Comment (mrIncompetent) [reply api/pkg/controller/container-linux/resources/clusterrole.go:13]: Oh yes.

**Comment by mrIncompetent** [comment]
> More of a sidequestion, does the operator drain and taint nodes prior to upgrading them?

Yes (Except it sets `node.spec.unschedulable=true` - So exactly what `kubectl drain does`)

**Comment by mrIncompetent** [comment]
/retest

@alvaroaleman PTAL

**Comment by mrIncompetent** [comment]
@alvaroaleman i screwed up the naming of the resources...

**Pull request merged**